feat: Adds SSL mutual auth support to intra-cluster requests

AlanConfluent opened the pull request:
### Description 
Allows for the configuration of SSL mutual auth for the internal listener `ksql.internal.listener`.  KSQL then considers the request to be done by a `SystemUser` and considers it authenticated.

This is done by supporting a separate key store and trust store in the server for internal requests, and using these when issuing inter-node requests.

This way, the main ssl configurations can be used for the public API (e.g. https://external.example.com) and new internal configs can be used to set the ssl configuration for each node (e.g. https://node-1.internal.example.com).  This way each node can have an internal identity, which is required to verify the certificates for internal requests.

### Testing done 
Wrote a functional test that uses https on both the listeners, each with a different certificate, as well as basic auth.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
````diff
@@ -219,6 +219,42 @@ credentials when starting the CLI by using the `--user` and
 ```
 
+Configure ksqlDB for Internal Authentication
+------------------------------------
````
```suggestion
--------------------------------------------
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,6 +219,42 @@ credentials when starting the CLI by using the `--user` and
+Configure ksqlDB for Internal Authentication
+------------------------------------
+
+Since some system requests aren't done on behalf of a user (e.g. `/heartbeat`, `/lag`), 
```
```suggestion
Since some system requests aren't done on behalf of a user, like `/heartbeat` and `/lag`, 
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,6 +219,42 @@ credentials when starting the CLI by using the `--user` and
+connection level, but ksqlDB then also considers these requests to be authenticated.
+
+
+You can read generally about configuring `ksql.internal.listener` [here](index.md).
```
```suggestion
For more information about configuring `ksql.internal.listener`, see [Configure ksqlDB for HTTPS](index.html#configure-ksqldb-for-https).
```

purplefox commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/api/auth/SystemAuthenticationHandlerTest.java:
```diff
@@ -0,0 +1,87 @@
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class SystemAuthenticationHandlerTest {
```
I'm not a fan of fine grained unit tests like these - they end up being very brittle and can provide a false sense of security. Would much rather see more coarse grained tests :)

purplefox replied:
It's a bit incongruous to add a fine grained mock test for this handler - the other handlers don't have this and that's been the approach we've taken with the new API. Instead we use much more coarse grained test which test the code as it would actually be used in a real setup. E.g. take a look at AuthTest.

purplefox replied:
Unlike full blown integration tests these kinds of tests run really fast, so you should be able to write enough tests that exercise all the code paths, but unlike fine grained unit tests, they run exactly as they would in the real system, and we're not testing the implementation directly thus allowing it to evolve and be refactored much more easily.

BTW... I'm well known for my opinion on fine grained unit tests (you should see my Twitter conversations), and I'm currently thinking of writing a blog post on them and why I think this approach - mainly championed by "TDD" has been super damaging in our industry ;)

AlanConfluent replied:
I was following the pattern in ksqldb-rest-app/src/test/java/io/confluent/ksql/api/auth/... where other auth handers do this kind of test.  It would seem a shame to delete my nice test.

I'm happen to try to write a few test cases in AuthTest to specifically test this, though it's likely going to be a bit complex since it requires setting up https.  I'll give this a shot.

AlanConfluent commented:
> SNI is new to me, but from my quick google it looks like SNI is related to virtual hosting, and seems completely unnecessary for ksqlDB. Have you enabled this to allow a ksqlDB node to use different ssl certs for its internal and external listeners? Do we think this is something users need?

From my understanding, it enables the serving of multiple different identities/certificates, depending on the host name requested.  It may get most use in virtual hosting, but my intention was to have both an external and an internal identity.  The reason has to do with how they're requested and verified.

Externally, some user may request https://external.example.com and the server returns a cert which has the name "external.example.com".  This match is what allows the client to verify the identity.

If internally, we have a different hostname, say https://internal.example.com, and it returns a cert with the name "external.example.com", verification is not possible.  With SNI enabled (with two identities in the key store), the server matches the requested hostname with certs in its store and is able to return one for "internal.example.com" instead.  Of course, we can avoid entirely using SNI by just configuring the internal listener with a separate key store with a cert for "internal.example.com".

> However, thinking this through, each ksqlDB node only needs one identity, so only one ssl cert, which can be used for either the client facing endpoint, the internal endpoint, or both. Likewise, only a single trust store would be needed, (or no trust store at all), which would include the CAs / certs the node should trust. Then, as you have done, we'd need two separate 'client auth mode' settings to control if mutual auth is required for the client facing and internal endpoints.


If we want just one identity, the externally facing identity, it makes the verification on the client more complex.  Also, it likely prevents you from having different identities for each node in your cluster (since you probably have just one externally facing cert). On the plus side, any KSQL currently configured to do https externally can just turn this on.  We have a couple options to do this:
- Route the internal requests externally using the external hostname, just as if it was an external client.  This breaks the notion of a network protected internal listener since that interface must now be exposed.  Also, any efficiencies that might have been gained may be lost since it would likely mean routing the request outside of the local network, and back in again.  Also, it might be hard to differentiate between a trusted, mutually authed client, and an internal cluster node.
- Not sure if you noticed the `HostAliasResolver` interface I added (for testing!), but it effectively allows you to make aliases for a hostname such that you could say I want to request host "external.example.com" using a `SocketAddress` referring to https://internal.example.com:internap_port.  This would allow requests to the internal address, but would compare the cert to "external.example.com".  The only downside is it adds complexity to the requests since there's a notion of recipient identity vs actual hostname.

Between these, the second is fairly straightforward to do.  It would only require configuring internal clients correctly.

Do you see why I had originally thought to introduce two identities?

> The other main point I'd like to raise is that this PR adds complexity to the internal routing of requests as you've only added support for inter-node mutual-auth for _some_ of the inter-node communication.
> All inter-node communication should be covered by the enhancement you're making! If that means moving proxied pull query requests or scattered show queries requests onto a internal specific endpoint so that we can separate them from client requests, then we should do that. It sounds like a good long term design to clearly separate the two anyway.

That's a fair comment.  I considered that I might do this in two parts to reduce complexity by only using this on system endpoints first.  But it's very easy to it for all internal calls on the internal listener.  Since proxied pull query requests or scattered show queries are exposed separately on the internal listener, they are already separated, even if not in code organization.

I had originally had two separate Verticles, one with external, and the other internal, but I was told by @purplefox that in Vert.x this is an anti-pattern, so we have one Verticle installed in two places.

> 
> Finally, there's a lot of code changes to existing integration tests in this PR. Why is this? I wouldn't have expected those tests to have to change at all. Is this change because of the SNI and/or the fact that we now need two clients for internal communication?

I believe that most of the changes I made were in the new test I added.  I added a bunch of helper functions that are necessary to do https requests in tests using `HostAliasResolver`.  These were to test with certain certs even though the requests are going to localhost.

AlanConfluent commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/api/auth/SystemAuthenticationHandlerTest.java:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  @Test
+  public void success() {
```
Returned the handler.  Improved the name.

AlanConfluent commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/HighAvailabilityTestUtil.java:
```diff
@@ -38,9 +40,29 @@
-      final RestResponse<ClusterStatusResponse> res = restClient.makeClusterStatusRequest();
+  static ClusterStatusResponse sendClusterStatusRequest(
+      final TestKsqlRestApp restApp) {
+    try (final KsqlRestClient restClient = restApp.buildKsqlClient()) {
```
I had previously added the additional version below which allowed for https calls to the client API, so I could use different certs for the public listener and private.  But this wasn't really what I was testing, so I removed all of these and just added a simple auth credential version.

AlanConfluent commented on ksqldb-test-util/src/main/java/io/confluent/ksql/test/util/secure/ClientTrustStore.java:
```diff
@@ -42,7 +43,48 @@
   private static final String TRUSTSTORE_PASSWORD = "password";
+
+  // Trust store containing two certs, for internal.example.com and external.example.com
+  private static final String BASE64_ENCODED_STORE_MULTIPLE =
```
Done, moved to `MultiNodeKeyStore`.

AlanConfluent commented on ksqldb-test-util/src/main/java/io/confluent/ksql/test/util/secure/ServerKeyStore.java:
```diff
@@ -88,7 +89,103 @@
   private static final String TRUSTSTORE_PASSWORD = "password";
+
+  // Keystore containing two certs, for external.example.com and internal.example.com
+  private static final String BASE64_ENCODED_STORE_MULTIPLE =
```
Done, moved to `MultiNodeTrustStore`.

AlanConfluent commented:
@big-andy-coates I've addressed your comments.  I switched from using SNI to just using a separate key store and trust store.  It's simpler that way and also lets me control which identity is being sent from the client, which the Vert.x client api didn't otherwise allow.

AlanConfluent commented:
> Hey Alan,
> 
> A couple of questions/comments:
> 
> 1. Some of our users will be running a cluster in a completely secured environment with no public access and they don't require any kind of auth - does this solution allow all the internal endpoints to be used without auth requirement? (This shouldn't be the default, but needs to be supported).

This works fine with no auth.  If you don't set `ksql.internal.ssl.client.authentication`, it defaults to `NONE` and will fall back on other auth used in the system, which is to say that if you have jaas or the security plugin, it will disallow internal endpoints (though you can explicitly whitelist them if you want).  If you have no auth configured, it will require no auth.

> 2. Do we now require executing inter node pull query requests over TLS? If so, have we you considered the performance implications? TLS handshake can have twice the network latency of a non TLS handshake. If this is the case we should consider prioritising reusing of connections for inter node pull query requests.

For forwarded pull queries, yes, it would be over TLS.  I do think that connection pooling would be required to enable if performance was a big concern.

AlanConfluent commented:
@big-andy-coates We had a conversation the other day offline and I think we're on the same page with respect to using two certificates.

Also, I was able to figure out how to use an alias to choose the certificate we want.  While Vert.x doesn't directly support it, I wrote a quick utility that loads the keystore in memory, grabs the cert/key with the given alias, and creates an in memory single-entry version that can be passed to Vert.x.  This way, we just have a config for the alias we want to use, if any and it all reads from the same keystore.  It's a better fit for us than SNI since that allows for serving different certs from the same listener, which we don't need.

purplefox commented:
> Also, I was able to figure out how to use an alias to choose the certificate we want. While Vert.x doesn't directly support it, I wrote a quick utility that loads the keystore in memory, grabs the cert/key with the given alias, and creates an in memory single-entry version that can be passed to Vert.x. This way, we just have a config for the alias we want to use, if any and it all reads from the same keystore. It's a better fit for us than SNI since that allows for serving different certs from the same listener, which we don't need.

Glad to hear that approach worked :)

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/services/DefaultKsqlClient.java:
```diff
@@ -164,5 +167,50 @@ private static HttpClientOptions createClientOptions() {
+      final String internalAlias = clientProps
+          .get(KsqlRestConfig.KSQL_SSL_KEYSTORE_ALIAS_INTERNAL_CONFIG);
+      final String trustStoreLocation = clientProps.get(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG);
+      if (trustStoreLocation != null) {
```
```suggestion
      if (trustStoreLocation != null && !trustStoreLocation.isEmpty()) {
```

???

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/services/DefaultKsqlClient.java:
```diff
@@ -164,5 +167,50 @@ private static HttpClientOptions createClientOptions() {
+            .setPassword(suppliedTruststorePassword == null ? "" : suppliedTruststorePassword));
+
+        final String keyStoreLocation = clientProps.get(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG);
+        if (keyStoreLocation != null) {
```
```suggestion
        if (keyStoreLocation != null && !keyStoreLocation.isEmpty()) {
```

??? 

PITA these configs don't default to "" and have a validator to ensure they're not null!

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/services/DefaultKsqlClient.java:
```diff
@@ -164,5 +167,50 @@ private static HttpClientOptions createClientOptions() {
+              .get(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG));
+          final JksOptions keyStoreOptions = new JksOptions()
+              .setPassword(suppliedKeyStorePassord);
+          if (internalAlias != null && !internalAlias.isEmpty()) {
```
```suggestion
          if (!internalAlias.isEmpty()) {
```

This one we control, so can have suitable defaults and validators to ensure never null.

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/services/DefaultKsqlClient.java:
```diff
@@ -164,5 +167,50 @@ private static HttpClientOptions createClientOptions() {
 
+  private static KsqlClient getInternalClient(final Map<String, String> clientProps,
+      final Optional<HostAliasResolver> hostAliasResolver) {
+    final boolean verifyHost =
+        !KsqlRestConfig.SSL_CLIENT_AUTHENTICATION_NONE.equals(clientProps.get(
+        KsqlRestConfig.KSQL_INTERNAL_SSL_CLIENT_AUTHENTICATION_CONFIG));
```
Does this work if mode is `REQUESTED`?

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/KeystoreUtil.java:
```diff
@@ -0,0 +1,103 @@
+      final String password,
+      final String alias
+  ) {
+    final char[] pw = password != null ? password.toCharArray() : null;
```
Better to encode the optionality of `password` into the signature of the method? i.e. `Optional<String> password`?

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/KeystoreUtil.java:
```diff
@@ -0,0 +1,103 @@
+      final String alias
+  ) {
+    final char[] pw = password != null ? password.toCharArray() : null;
+    final KeyStore keyStore = loadExistingKeyStore(keyStorePath, pw);
+
+    final PrivateKey key;
+    final Certificate[] chain;
+    try {
+      key = (PrivateKey) keyStore.getKey(alias, pw);
+      chain = keyStore.getCertificateChain(alias);
```
Is this assuming the same password for the keystore and the key?  Isn't it possible these are different?

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/KeystoreUtil.java:
```diff
@@ -0,0 +1,103 @@
+      key = (PrivateKey) keyStore.getKey(alias, pw);
+      chain = keyStore.getCertificateChain(alias);
+    } catch (Exception e) {
+      throw new KsqlException("Error fetching key/certificate", e);
```
Include the `alias` in the message to make it clear _which_ is failing?

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/KeystoreUtil.java:
```diff
@@ -0,0 +1,103 @@
+
+  private static KeyStore loadExistingKeyStore(final String keyStorePath, final char[] pw) {
+    FileInputStream input = null;
+    try {
```
Why not use try-with-resource, removing the need for finally block

```suggestion
    try (FileInputStream input = new InputStream(keyStoreParth) {
```

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/KeystoreUtil.java:
```diff
@@ -0,0 +1,103 @@
+    try {
+      input = new FileInputStream(keyStorePath);
+      final KeyStore keyStore = KeyStore.getInstance(KEYSTORE_TYPE);
+      keyStore.load(new FileInputStream(keyStorePath), pw);
```
input stream leak...

```suggestion
      keyStore.load(input, pw);
```

big-andy-coates commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlClient.java:
```diff
@@ -44,27 +45,50 @@
+      final Optional<BasicCredentials> credentials,
+      final LocalProperties localProperties,
+      final HttpClientOptions httpClientOptions,
+      final Consumer<HttpClientOptions> sslHttpClientOptionsConsumer,
```
why are we passing in a consumer that is just called back with the `httpClientOptions` passed in?  This feels like an anti-pattern to me.

big-andy-coates commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlClient.java:
```diff
@@ -44,27 +45,50 @@
   public KsqlTarget target(final URI server) {
     final boolean isUriTls = server.getScheme().equalsIgnoreCase("https");
     final HttpClient client = isUriTls ? httpTlsClient : httpNonTlsClient;
+    final String aliasHost = hostAliasResolver.map(resolver -> resolver.resolve(server.getHost()))
+        .orElse(server.getHost());
```
Rather than store a `Optional<HostResolver>`, why not default to a `HostResolver` impl that just returns the value passed? e.g. `Function.identify()`.  This would simplify this code.

big-andy-coates commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlClient.java:
```diff
@@ -44,27 +45,50 @@
       final LocalProperties localProperties,
-      final HttpClientOptions httpClientOptions
+      final HttpClientOptions httpClientOptions,
+      final Optional<HostAliasResolver> hostAliasResolver
```
It seems a shame to require every use of this _production_ class to supply a `HostAliasResolver`, given this should only be passed for _test_ cases, right?  And most test cases won't need this either. Just one or two specific tests that are testing the new code you're adding. Yet `HostAliasResolver` is mentioned around 40 times in this PR.

Is there another way?  

At the very least you could remove this parameter from this constructor and add another constructor, annotated with `@VisibleForTesting` that takes the resolver.  However, ideally, the production code should know nothing of `HostAliasResolver`...

I think this can be achieved by injecting an interface to handle the `SocketAddress.inetSocketAddress` call.   So rather than taking a `HostAliasResolver` the `@VisiableForTesting` constructor takes `BiFunction<Integer, String, SocketAddress> socketAddressFactory`.  Test code can then use this to perform the same aliasing that `HostAliasResolver` does now.   Please see if you can remove `HostAliasResolver` from the production code base, or at least from the non- `@VisibleForTesting` constructors.




SocketAddress.inetSocketAddress(server.getPort(), aliasHost)

big-andy-coates commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTarget.java:
```diff
@@ -73,28 +73,31 @@
       final LocalProperties localProperties,
-      final Optional<String> authHeader
+      final Optional<String> authHeader,
+      final String host
```
why are we passing in `host` here? I'm assuming this is different to `socketAddress.getHost`.  This is really counter-intuitive and the kind of thing that will trip people up later when they make changes to this class.

Is there any reason by you can't just change the calling code to pass in the _right_ `socketAdrress`?

big-andy-coates replied:
Looks good.

AlanConfluent commented:
@big-andy-coates You had wanted another review after the last.  Please tell me if you're good with it.

big-andy-coates commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,6 +219,134 @@ credentials when starting the CLI by using the `--user` and
+ksql.internal.listener=https://node-1.internal.example.com:8099
+
+# This enables mutual auth checking for the internal listener
+ksql.internal.ssl.client.authentication=REQUIRED
```
Maybe include `ksql.ssl.client.authentication=` whatever, just so users are aware?

big-andy-coates commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,6 +219,134 @@ credentials when starting the CLI by using the `--user` and
+different from its internal identity.  In order to create such a key store,
+refer [below](#setting-up-a-key-store-and-trust-store).
+
+In such a configuration, you must specify which key pair is used for a given
+listener by providing a key store alias.  For example, 
+if set, `ksql.ssl.keystore.alias.internal` will be used to find the key store entry
+with the given alias when setting up the internal listener.  Similarly,
+`ksql.ssl.keystore.alias.external` is used for the client listener `listeners`.
+Below is an example configuration:
```
perfect!

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/Server.java:
```diff
@@ -251,30 +255,57 @@ private static HttpServerOptions createHttpServerOptions(final KsqlRestConfig ks
-      if (keyStorePath != null && !keyStorePath.isEmpty()) {
-        options.setKeyStoreOptions(
-            new JksOptions().setPath(keyStorePath).setPassword(keyStorePassword.value()));
+      if (isInternalListener) {
+        final String internalAlias = ksqlRestConfig
+            .getString(KsqlRestConfig.KSQL_SSL_KEYSTORE_ALIAS_INTERNAL_CONFIG);
+        setTlsOptions(ksqlRestConfig, options, internalAlias,
+            ksqlRestConfig.getClientAuthInternal());
+      } else {
+        final String externalAlias = ksqlRestConfig
+            .getString(KsqlRestConfig.KSQL_SSL_KEYSTORE_ALIAS_EXTERNAL_CONFIG);
+        setTlsOptions(ksqlRestConfig, options, externalAlias, ksqlRestConfig.getClientAuth());
```
Just a suggestion but:

```suggestion
      final String ksConfigName = isInternalListener:
           ? KsqlRestConfig.KSQL_SSL_KEYSTORE_ALIAS_INTERNAL_CONFIG
           : KsqlRestConfig.KSQL_SSL_KEYSTORE_ALIAS_EXTERNAL_CONFIG;
           
      final String alias = ksqlRestConfig.getString(ksConfigName);
      setTlsOptions(ksqlRestConfig, options, alias, ksqlRestConfig.getClientAuthInternal());
```

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java:
```diff
@@ -364,6 +384,27 @@
+        ).define(
+            KSQL_SSL_KEYSTORE_ALIAS_EXTERNAL_CONFIG,
+            Type.STRING,
+            "",
```
cool

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/Server.java:
```diff
@@ -251,30 +254,54 @@ private static HttpServerOptions createHttpServerOptions(final KsqlRestConfig ks
+    if (keyStorePath != null && !keyStorePath.isEmpty()) {
+      final JksOptions keyStoreOptions = new JksOptions()
+          .setPassword(keyStorePassword.value());
+      if (keyStoreAlias != null && !keyStoreAlias.isEmpty()) {
```
Not sure. Maybe explicitly setting it via `ksql.what.ever=null` in the config file?  If that doesn't allow nulls then great!

big-andy-coates commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/SystemAuthenticationFunctionalTest.java:
```diff
@@ -0,0 +1,331 @@
+    private static final TestKsqlRestApp REST_APP_0 = TestKsqlRestApp
+        .builder(TEST_HARNESS::kafkaBootstrapServers)
+        .withEnabledKsqlClient(LOCALHOST_FACTORY)
+        .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://0.0.0.0:8088")
```
Avoid hardcoded ports in tests - it causes tests to fail if the ports in use, e.g. 
* If the build server is running two builds on the same box
* An engineer is running two builds, for two different change sets, at the same time.
* An engineer is running ksql to test something else, while they've also got a build running.
* etc.

I *think* you should be able to just set it to `http://0.0.0.0:0` and the server will auto-pick a free port.  It certainly did in Jetty days, not sure about Vert.x.  Check with Tim if it doesn't work. I'm sure he'll know how to.

Please update all places the port is hardcoded.

big-andy-coates commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/services/TestDefaultKsqlClientFactory.java:
```diff
@@ -11,8 +13,9 @@
-  public static SimpleKsqlClient instance(Map<String, Object> clientProps) {
-    return new DefaultKsqlClient(Optional.empty(), clientProps);
+  public static SimpleKsqlClient instance(Map<String, Object> clientProps,
+      final BiFunction<Integer, String, SocketAddress> socketAddressFactory) {
```
Is it worth having two versions: one with and one without this function?  Seems like most things won't need the ability to override the host.

big-andy-coates commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlClient.java:
```diff
@@ -55,16 +58,33 @@ public KsqlClient(
     this.httpTlsClient = createHttpClient(vertx, clientProps, httpClientOptions, true);
   }
 
+  public KsqlClient(
```
Worth adding java docs on the params?  The new `socketAddressFactory` probably isn't obvious

AlanConfluent commented on docs/operate-and-deploy/installation/server-config/security.md:
````diff
@@ -219,6 +219,134 @@ credentials when starting the CLI by using the `--user` and
 <ksql-install>bin/ksql --user fred --password letmein http://localhost:8088
 ```
 
+Configure ksqlDB for Internal Authentication
````
> I've just had a look at what's already in this document and I think, rather than adding a brand new section, the details you're covering need to be integrated with what is already there.

I followed your instructions fairly closely, so I think we're good there.

@JimGalasyn Can you take a look at my docs changes again?  Also, How can I verify that the local anchor links work correctly?  None of my tools allow for checking this.

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -6,21 +6,39 @@ description: Settings for security ksqlDB
-  such SSL for encryption and mutual authentication for authorization.
-- ksqlDB supports SSL on all network traffic.
+ksqlDB supports several combinations of encryption and authentication on its
+client-facing and internal endpoints, and also supports many of the security
```
```suggestion
client-facing and internal endpoints. ksqlDB also supports many of the security
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -6,21 +6,39 @@ description: Settings for security ksqlDB
+--------------------------
+
+This section covers how to secure installations of ksqlDB outside of Confluent Cloud,
+e.g. on-premise installations or manual installations on other cloud platforms.
```
```suggestion
like on-premises installations or manual installations on other cloud platforms.
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -36,8 +54,75 @@ configuration file specified.
+Securing interactive deployments
+--------------------------------
+
+Securing the interactive ksqlDB installation involves securing the HTTP endpoints the ksqlDB server is listening on.
```
```suggestion
Securing the interactive ksqlDB installation involves securing the HTTP endpoints that the ksqlDB server is listening on.
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -36,8 +54,75 @@ configuration file specified.
+You may still have to [secure communication](#securing-communication-with-other-services) with other services.
+
+
+Securing single listener setup
```
```suggestion
Securing single-listener setup
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
````diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
 ```
 
-Configure ksqlDB for Confluent Cloud
+Securing dual listener setup
````
```suggestion
Securing dual-listener setup
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+
+Using dual listeners for ksqlDB is appropriate when the client and
+inter-node communication utilize different authentication and security
+configurations.  This is most likely the case when ksqlDB is deployed as an
```
```suggestion
configurations. This is most likely the case when ksqlDB is deployed as an
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+communication and uses basic authentication for the external client API.
+This is most likely to be paired with SSL above.
+
+Configuring internal for SSL-mutual authentication
```
"internal" doesn't seem quite right here.

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
````diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+ksql.internal.ssl.client.authentication=REQUIRED
+```
+
+Configuring internal for SSL-mutual authentication + external for HTTP-BASIC authentication
````
```suggestion
Configuring internal for SSL-mutual authentication and external for HTTP-BASIC authentication
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
````diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+```
+
+Configuring internal for SSL-mutual authentication + external for HTTP-BASIC authentication
+--------------------------------------------------------------------------------
````
```suggestion
---------------------------------------------------------------------------------------------
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+Configuring internal for SSL-mutual authentication + external for HTTP-BASIC authentication
+--------------------------------------------------------------------------------
+
+Client facing basic HTTP authentication can be used alongside authentication for the
```
```suggestion
Client-facing basic HTTP authentication can be used alongside authentication for the
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+For more detail on basic authentication,
+[see above](#configure-ksqldb-for-basic-http-authentication).
+
+Configuring internal for SSL-mutual authentication + external for SSL encryption
```
```suggestion
Configuring internal for SSL-mutual authentication and external for SSL encryption
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+[see above](#configure-ksqldb-for-basic-http-authentication).
+
+Configuring internal for SSL-mutual authentication + external for SSL encryption
+--------------------------------------------------------------------------------
```
```suggestion
------------------------------------------------------------------------------------
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+If you want to use HTTPS on `listeners` as well as use SSL mutual
+auth for internal communication on `ksql.internal.listener`, you will likely
+require two different key pairs, since your host's identity to clients may be
+different from its internal identity.  In order to create such a key store,
```
```suggestion
different from its internal identity. To create such a key store,
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+refer [below](#setting-up-a-key-store-and-trust-store).
+
+In such a configuration, you must specify which key pair is used for a given
+listener by providing a key store alias.  For example,
```
```suggestion
listener by providing a key store alias.  For example,
```
```suggestion
listener by providing a key store alias. For example,
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/security.md:
```diff
@@ -219,13 +304,193 @@ credentials when starting the CLI by using the `--user` and
+In such a configuration, you must specify which key pair is used for a given
+listener by providing a key store alias.  For example,
+if set, `ksql.ssl.keystore.alias.internal` will be used to find the key store entry
+with the given alias when setting up the internal listener.  Similarly,
```
```suggestion
with the given alias when setting up the internal listener. Similarly,
```
